types/ramda/test/assoc-tests.ts

```diff
@@ -80,9 +80,3 @@ import * as R from "ramda";
     // $ExpectType Record<"c", string> & Omit<{ a: number; b: number; c: number; }, "c">
     R.assoc("c")("test", { a: 1, b: 2, c: 3 });
 });
-
-(() => {
-    type ABC = Record<string, string>;
-    const b: ABC = R.compose(R.assoc, R.toString)(3)("c", { 1: "a", 2: "b" }); // => {1: "a", 2: "b", 3: "c"}
```

Comment:
I was able to fix this by moving the first overload of `assoc` (`assoc<K extends string>(prop: K)`) to the end of the overload list.

Reply:
The overload order change for `assoc` was done in https://github.com/ramda/types/pull/96. This re-order was intentional, as the previous ordering was incorrect. And therefore the function signature returned by `R.compose(R.assoc, R.toString)`

This was to match how [typescript wants you to overload](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads). The full signature needs to come last.

Previously, `assoc(key, value, obj)` was not the bottom overload, but should have been. This test previously worked because it was not

I could have updated this test to be this and it would have worked as well
```typescript
const b: ABC = R.compose(R.assoc, R.toString)(3, "c", { 1: "a", 2: "b" }); // => {1: "a", 2: "b", 3: "c"}
```
I choose to remove the test though because all the overloads for `assoc` are thoroughly tested types-ramda

It is true that this could break some existing code, but in a net-positive way, since the resultant required change is correct and matches the behavior of all other ramda functions used in this manner.

Reply:
> The full signature needs to come last.

The implementation signature certainly needs to come last, but that isn't counted in the overload list that appears to callers. When two signature argument lists are incompatible with each other, their order doesn't matter. When two or more signature argument lists are compatible, you want the narrower argument lists to come first.

There are some complexities around how overload lists are handled when a function is used as a callback, though, as TypeScript allows you to oversupply arguments to a callback so that you can write `array.map(value => ...)` without needing to specify parameters for each unused argument. As a result, a function that accepts a `(x: T) => U` callback can be satisfied by any overload with at least one argument. This is further complicated by the fact that, while function calls bind the first overload that matches the argument list, callbacks (and function type inference) don't always take the first overload if there is a later one that *seems* like a better fit.

If it is expected that `R.assoc` will be treated most often as a unary function when used as a callback, it might make sense to prefer the unary argument list by placing it last in the overload list. If not, then I wouldn't change it.

Reply:
So long as it doesn't break your other test cases, you could also try duplicating the `<K extends string>(prop: K)` overload so that it is both the first *and* the last overload in the list. It's duplicative, but gets around TypeScript's overload resolution picking the higher-arity overload.

Reply:
> TypeScript's overload resolution picking the higher-arity overload.

I _want_ it to pick the higher-arity overload. All other ramda functions work this way